### PR TITLE
WIP: TASK: Refactor and cleanup Neos.Neos FusionView

### PR DESCRIPTION
**Pull request by mhsdesign**
**Upgrade instructions**

**Review instructions**

Related https://neos-project.slack.com/archives/C3MCBK6S2/p1694339628529459?thread_ts=1694269188.257739&channel=C3MCBK6S2&message_ts=1694339628.529459

Im just currently looking into wether i could fetch documentNode and siteNode with one query via

```
$ancestors = $subgraph->findAncestorNodes($node->nodeAggregateId, FindAncestorNodesFilter::create(nodeTypeConstraints: 'Neos.Neos:Document');

$documentNode = $ancestors->first();

$siteNode = $ancestors->last();
```

but i dont know if its a criteria for the homepage node to be of type Neos.Neos:Document?

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Classes/View/FusionView.php [hunk 1]
```diff
@@ -99,161 +47,134 @@ public function render(): string|ResponseInterface
...
-        $fusionRuntime = $this->getFusionRuntime($currentSiteNode);
+        // clear the cached runtime instance
+        unset($this->fusionRuntime);
 
-        return $fusionRuntime->canRender($this->fusionPath);
+        $this->fusionRenderingStuff = $this->renderingUtility->createFusionRenderingStuff(
+            $value,
+            $this->controllerContext->getRequest()
+        );
+        return $this;
```
Comment (mhsdesign) [inline Neos.Neos/Classes/View/FusionView.php:99]: The biggest difference user wise is, that "value" will be eagerly resolved to "node" "documentNode" and etc. 

This is a mini performance gain, so that `canRenderWithNodeAndPath` and `render` dont have to each recompute the values.

This change is only noticeable when "value" is not a valid node. As exceptions are now thrown already in the "assign" step and not before hand.

**Comment by mhsdesign** [comment]
Will also provide a new testcase for the Fusion view - currently it is disabled.

Edit: Rewrite for FusionViewTest done ;)

File: Neos.Neos/Classes/Domain/Service/RenderingUtility.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\Flow\Mvc\ActionRequest;
+use Neos\Fusion\Core\FusionGlobals;
+use Neos\Neos\Domain\Model\FusionRenderingStuff;
+use Neos\Neos\Domain\Model\RenderingContext;
+use Neos\Neos\Domain\Repository\SiteRepository;
+use Neos\Flow\Annotations as Flow;
+
+#[Flow\Scope('singleton')]
+class RenderingUtility
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Domain/Service/RenderingUtility.php:18]: Those should better be moved in the via https://github.com/neos/neos-development-collection/pull/4496 "emptied" `FusionService`

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php [hunk 3]
```diff
@@ -104,6 +104,16 @@ public function first(): ?Node
         return null;
     }
 
+    public function last(): ?Node
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php:107]: Seems like this should be a separate PR?

File: Neos.Neos/Classes/Domain/Model/FusionRenderingStuff.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Neos\Domain\Model;
+
+use Neos\Fusion\Core\FusionGlobals;
+
+final readonly class FusionRenderingStuff
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Domain/Model/FusionRenderingStuff.php:9]: we dislike this naming

Comment (mhsdesign) [reply Neos.Neos/Classes/Domain/Model/FusionRenderingStuff.php:9]: @kitsunet do you think a model like this makes sense?

I think yes, as those are all the ingredients/recipe for a sucessfull neos node rendering

File: Neos.Neos/Classes/View/FusionView.php [hunk 5]
```diff
@@ -99,161 +44,128 @@ public function render(): string|ResponseInterface
...
 
-    /**
-     * @return string
-     */
-    public function getFusionPath()
-    {
-        return $this->fusionPath;
-    }
+        $this->fusionRuntime->pushContextArray(
+            $this->fusionRenderingStuff->fusionContext->toContextArray()
```
Comment (kitsunet) [inline Neos.Neos/Classes/View/FusionView.php:114]: I would actually prefer to make this more flexible than before. Have `assign` work as intended and add the other assigned variables to the context array here.

**Comment by mhsdesign** [comment]
When revisiting this topic (see https://github.com/neos/neos-development-collection/issues/4933) ill probably be able to cherrypick some parts from it or get some inspiration from my older self but this pr can be closed.

Many things i was trying to improve have also been adjusted in the meantime:
- The HTTP message decoding will be handled from the runtime
- addition of `Nodes::last`
- introduction of `Subgraph::findClosest(Document)`
- introduction of Site NodeType


Also the rewrite of the Neos.Neos/Tests/Unit/View/FusionViewTest.php functional test turned out to be a bad idea, well behat test the fusion view ^^
